### ISSUE TITLE
Type inference does not detect missing value in certain data sets

### ISSUE THREAD

**Issue by iqml** [JuliaData/CSV.jl#390]
With CSV.jl `v0.4.3`, DataFrames.jl `v0.17.1` and Julia `v1.1`, the following code demonstrates the problem:

```jl
using CSV: File, write
using DataFrames: DataFrame

BEFORE = 9_900
AFTER = 101
FILEPATH = "baddata.csv"

# Make the CSV file
a = repeat(Union{Missing, Int64}[0], BEFORE)
push!(a, missing)
append!(a, repeat(Union{Missing, Int64}[0], AFTER))
DataFrame(a=a) |> write(FILEPATH)

# Get the column type using CSV.File
coltype = File(FILEPATH; allowmissing = :auto).types[1]
@info "Column type:" coltype
@assert coltype == Union{Missing, Int64}
```

The code writes a CSV file with one column, called `a`. The column contains:

1. 9,900 rows of the integer `0` (defined as `BEFORE`),
2. then a missing value,
3. then 101 rows of the integer `0` (defined as `AFTER`).

This is the smallest and simplest dataset I could identify that causes the problem. I have also run it on several different machines to verify.

The type of the column `a` should be `Union{Missing, Int64}`. However, loading the file with `CSV.File()` with `allowmissing = :auto` declares that the column type is `Int64`.

Note that if `BEFORE` is less than 9,900 or if `AFTER` is less than 101, the type is correctly identified and the error does not occur.

For example, if `BEFORE` is 1,000,000 but `AFTER` is only 100 or less, the type is correctly identified and the error does not occur.

So this works:

```jl
using CSV: File, write
using DataFrames: DataFrame

BEFORE = 1_000_000
AFTER = 100
FILEPATH = "baddata.csv"

# Make the CSV file
a = repeat(Union{Missing, Int64}[0], BEFORE)
push!(a, missing)
append!(a, repeat(Union{Missing, Int64}[0], AFTER))
DataFrame(a=a) |> write(FILEPATH)

# Get the column type using CSV.File
coltype = File(FILEPATH; allowmissing = :auto).types[1]
@info "Column type:" coltype
@assert coltype == Union{Missing, Int64}
```

But this does not:

```jl
using CSV: File, write
using DataFrames: DataFrame

BEFORE = 1_000_000
AFTER = 101
FILEPATH = "baddata.csv"

# Make the CSV file
a = repeat(Union{Missing, Int64}[0], BEFORE)
push!(a, missing)
append!(a, repeat(Union{Missing, Int64}[0], AFTER))
DataFrame(a=a) |> write(FILEPATH)

# Get the column type using CSV.File
coltype = File(FILEPATH; allowmissing = :auto).types[1]
@info "Column type:" coltype
@assert coltype == Union{Missing, Int64}
```

So both `BEFORE` must be greater than or equal to 9,900 AND `AFTER` must be greater than or equal to 101 for the error to occur.

Thanks.

**Comment by nalimilan** [comment]
Unfortunately that's expected as we don't parse all rows to detect type (which would kill performance). That's why the default is not `:auto`. See https://github.com/JuliaData/CSV.jl/pull/203 for a mitigation strategy.

**Comment by quinnj** [comment]
This will be fixed in #395

**Comment by quinnj** [comment]
Fixed in #395
